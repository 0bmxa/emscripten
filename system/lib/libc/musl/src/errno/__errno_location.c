#include "pthread_impl.h"

int *__errno_location(void)
{
	static int e;
	if (libc.main_thread) return pthread_self()->errno_ptr;
	return &e;
}
