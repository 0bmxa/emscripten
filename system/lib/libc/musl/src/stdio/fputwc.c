#include "stdio_impl.h"
#include <wchar.h>
#include <limits.h>
#include <ctype.h>
#include <string.h> // XXX EMSCRIPTEN

wint_t __fputwc_unlocked(wchar_t c, FILE *f)
{
	char mbc[MB_LEN_MAX];
	int l;

#if 0 // XXX EMSCRIPTEN
	f->mode |= f->mode+1;

	if (isascii(c)) {
		c = putc_unlocked(c, f);
	} else if (f->wpos + MB_LEN_MAX < f->wend) {
		l = wctomb((void *)f->wpos, c);
		if (l < 0) c = WEOF;
		else f->wpos += l;
	} else {
		l = wctomb(mbc, c);
		if (l < 0 || __fwritex((void *)mbc, l, f) < l) c = WEOF;
	}
#else
    int i;

  memset(mbc, 0, sizeof(mbc));
	l = wctomb(mbc, c);
  if (l < 0) c = WEOF;
  else if (f != stdout && f != stderr) { // XXX
	  if (l < 4) l = 4;
  }
  for (int i = 0; i < l; i++) fputc(mbc[i], f);
#endif
	return c;
}

wint_t fputwc(wchar_t c, FILE *f)
{
	FLOCK(f);
	c = __fputwc_unlocked(c, f);
	FUNLOCK(f);
	return c;
}

weak_alias(__fputwc_unlocked, fputwc_unlocked);
weak_alias(__fputwc_unlocked, putwc_unlocked);
