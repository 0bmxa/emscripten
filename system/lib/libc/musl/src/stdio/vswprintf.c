#include "stdio_impl.h"
#include <limits.h>
#include <string.h>
#include <errno.h>
#include <stdint.h>
#include <wchar.h>

struct cookie {
	wchar_t *ws;
	size_t l;
};

static size_t sw_write(FILE *f, const unsigned char *s, size_t l)
{
	size_t l0 = l;
	int i = 0;
	struct cookie *c = f->cookie;
	if (s!=f->wbase && sw_write(f, f->wbase, f->wpos-f->wbase)==-1)
		return -1;
	while (c->l && l && (i=mbtowc(c->ws, (void *)s, l))>=0) {
		s+=i;
		l-=i;
		c->l--;
		c->ws++;
	}
	*c->ws = 0;
	return i<0 ? i : l0;
}

int vswprintf(wchar_t *restrict s, size_t n, const wchar_t *restrict fmt, va_list ap)
{
#if 0 // XXX EMSCRIPTEN
	int r;
	FILE f;
	unsigned char buf[256];
	struct cookie c = { s, n-1 };

	memset(&f, 0, sizeof(FILE));
	f.lbf = EOF;
	f.write = sw_write;
	f.buf_size = sizeof buf;
	f.buf = buf;
	f.lock = -1;
	f.cookie = &c;
	if (!n) {
		return -1;
	} else if (n > INT_MAX) {
		errno = EOVERFLOW;
		return -1;
	}
	r = vfwprintf(&f, fmt, ap);
	sw_write(&f, 0, 0);
	return r>=n ? -1 : r;
#else
  // XXX EMSCRIPTEN: use memfs through libc fs
  #define TEMPFILE "emscripten.vswprintf.temp.buffer"
  FILE *f = fopen(TEMPFILE, "wb");
  int r = vfwprintf(f, fmt, ap);
  fclose(f);
  f = fopen(TEMPFILE, "rb");
  fread(s, 1, (r>=n ? n : r)*4, f);
  fclose(f);
  remove(TEMPFILE);
  return r>=n ? -1 : r;
#endif
}
