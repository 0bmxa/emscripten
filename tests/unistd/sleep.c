#include <stdio.h>
#include <errno.h>
#include <unistd.h>
#include <time.h>

int main() {
  time_t start;

  start = time(0);
  printf("sleep(2) ret: %d\n", sleep(2));
  printf("after sleep(2): %d\n", time(0) - start);
  printf("errno: %d\n", errno);
  errno = 0;

  start = time(0);
  printf("usleep(3000000) ret: %d\n", usleep(3100000));
  printf("after usleep(3000000): %d\n", time(0) - start);
  printf("errno: %d\n", errno);

  return 0;
}
