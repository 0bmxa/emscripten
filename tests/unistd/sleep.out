sleep(2) ret: 0
after sleep(2): 2
errno: 0
usleep(3000000) ret: 0
after usleep(3000000): 3
errno: 0
